src/solid/lib/gate.ts

```diff
@@ -0,0 +1,98 @@
+import { Gate } from "../index.h";
+import { createEvent, createStore, Domain, launch } from "effector";
+import { onCleanup, onMount, createEffect } from "solid-js";
+
+export function useGate<Props>(
+  GateComponent: Gate<Props>,
+  props: Props = {} as any,
+) {
+  onMount(() => {
+    GateComponent.open(props);
+
+    onCleanup(() => GateComponent.close(props))
+  });
+
+  createEffect(() => {
+    // read every getter in props to subscribe
+    for (const _ of Object.values(props)) {}
+    GateComponent.set(props)
+  })
+}
+
+export function createGateImplementation<State>({
+  name = 'gate',
+  domain,
+  defaultState,
+  hook: useGateHook,
+}: {
+  name?: string
+  domain?: Domain
+  defaultState: State | {}
+  hook: typeof useGate
+}): Gate<State> {
+  const fullName = `${domain ? `${domain.compositeName.fullName}/` : ''}${name}`
+  const set = createEvent<State>(`${fullName}.set`)
+  const open = createEvent<State>(`${fullName}.open`)
+  const close = createEvent<State>(`${fullName}.close`)
+  const status = createStore(Boolean(false), {
+    name: `${fullName}.status`,
+  })
+    .on(open, () => Boolean(true))
+    .on(close, () => Boolean(false))
+  const state = createStore(defaultState as State, {
+    name: `${fullName}.state`,
+  })
+    .on(set, (_, state) => Object.create(
+      Object.getPrototypeOf(state),
+      Object.getOwnPropertyDescriptors(state)
+    ))
+    .reset(close)
+  if (domain) {
+    const {hooks} = domain as any
+    launch({
+      target: [
+        hooks.store,
+        hooks.store,
+        hooks.event,
+        hooks.event,
+        hooks.event,
+      ] as any,
+      params: [status, state, open, close, set],
+    })
+  }
+  function GateComponent(props: State) {
+    useGateHook(GateComponent as any, props)
+    return null
+  }
+  GateComponent.open = open
+  GateComponent.close = close
+  GateComponent.status = status
+  GateComponent.state = state
+  GateComponent.set = set
+  GateComponent.displayName = `Gate:${fullName}`;
+
+  return GateComponent;
+}
+
+export function createGate<Props>(
+  name: string = 'gate',
+  defaultState: Props = {} as any,
+): Gate<Props> {
```

Comment:
To be honest I was thinking about making Gate a bit different for solid.

There is rarely a need to check for all props.

I was thinking if something like this can be done:

```ts
import {createGate} from 'effector-solid'
import {render} from 'solid-js/web'

const Gate = createGate('gate with props')

const App = () => (
  <section>
    <Gate foo="bar" />
  </section>
)

Gate.state.foo.watch(foo => {
  console.log('current state of foo prop', foo)
})

// => current state of foo prop undefined

const unmount = render(() => <App />, document.getElementById('root'))
// => current state of foo prop bar
```

But this will probably have to involve proxies and will introduce undefined behaviour when used like this

```ts
// @ts-ignore
Gate.state.thisPropWillNeverExist.watch(() => {})
```

Reply:
I think, it actually can work, if gate state shape will be defined explicitly 🤔 
```ts
const Gate = createGate({
 foo: null,
 propB: "default"
})

const App = () => (
  <section>
    <Gate foo="bar" />
  </section>
)

// only props, that were initially setup at the Gate, will be available on state
Gate.state.foo.watch(foo => {
  console.log('current state of foo prop', foo)
})

// error immideatly
// not defined in default gate shape - not available at `state`
Gate.state.thisPropWillNeverExist.watch(() => {})
```

Reply:
Hmm, this might work, yeah.